proto/prysm/storage/beacon_storage.proto

```diff
@@ -0,0 +1,29 @@
+// Copyright 2021 Prysmatic Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+syntax = "proto3";
+
+package ethereum.eth.storage;
+
+import "proto/eth/v2/version.proto";
+
+
+// BeaconStateForStorage is used only inside db/kv package and it there for the sole purpose of
+// storing the state objects efficiently. With normalizing the fields in the "state" bucket
+// it also stores some indexes and other metadata fields that cannot be stored in the
+// BeaconState object itself.
+message BeaconStateForStorage {
+  v2.Version version = 1001;        // Used to identify the state object
+  bytes validatorIndexes = 1002;    // stores the compressed indexes of the validator entry hashes
```

Comment:
I think an outer wrapper with a oneOf could be a good move for future proofing. This would allow us to iterate on the schema of the inner BeaconState and provide value in migrations. For instance, we may change our minds on the design of validatorIndexes, or add additional fields. On that note I would also recommend leaving out validatorIndexes until that implementation is final. So more concretely here is what I'm imagining:
```
message BeaconStateForStorage {
    oneOf state_storage {
        BeaconStateVersionedBytes versioned_bytes = 1;
    }
}
message BeaconStateVersionedBytes {
    v2.Version version = 1;
    bytes state = 2;
}
```
cc @prestonvanloon what do you think?

Reply:
we wanted to keep state as ssz bytes since it was faster to unmarshall.
my original design was to have one of the three state objects.

Reply:
Right, in my example `bytes state = 2;` is still ssz bytes. The idea of my suggestion is to put the entire envelope into storage so we don't have a situation where some aspects of the schema are versioned and others aren't.

Reply:
> so we don't have a situation where some aspects of the schema are versioned and others aren't.

For some time, we cannot avoid people having different types of schema (older one, historic state, the new one).
We can offer them to move to "new schema" with a migrate subcommand.
When we go to Prysm 3.0 (merge),  we will not support older types of schema and support only one type.

Reply:
That's a good point, will the new storage format only be used after an offline migration has been performed, or does the DB code need to be able to differentiate between the 2 kinds of values somehow? If the latter, do you have a plan for dealing with that?

Reply:
It's the later. You can find out by looking at the migration bucket (which says what migration had been done, used already in code today).